runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -235,32 +258,50 @@ public int appendObjectState(Stream<Map.Entry> entryStream) {
         ImmutableMap<CheckpointEntry.CheckpointDictKey, String> mdkv =
                 ImmutableMap.copyOf(this.mdkv);
 
-        final Iterable<List<Map.Entry>> partitions =
-                Iterables.partition(entryStream::iterator, batchSize);
-        int entryCount = 0;
-
-        for (List<Map.Entry> partition : partitions) {
-            MultiSMREntry smrEntries = new MultiSMREntry();
-            for (Map.Entry entry : partition) {
-                smrEntries.addTo(new SMREntry("put",
-                        new Object[]{keyMutator.apply(entry.getKey()),
-                                valueMutator.apply(entry.getValue())},
-                        serializer));
-                entryCount++;
+        int totalEntryCount = 0;
+        int currentBatchSize = 0;
+        int numBytesPerCheckpointEntry = 0;
+        int maxWriteSizeLimit = (int) (BATCH_THRESHOLD_PERCENTAGE * rt.getParameters().getMaxWriteSize());
+        ByteBuf b = Unpooled.buffer();
+        MultiSMREntry smrEntries = new MultiSMREntry();
+
+        for (Map.Entry entry : (Iterable<Map.Entry>) entryStream::iterator) {
+            SMREntry each = new SMREntry("put",
+                    new Object[]{keyMutator.apply(entry.getKey()),
+                            valueMutator.apply(entry.getValue())},
+                    serializer);
+
+            each.serialize(b);
+            numBytesPerCheckpointEntry += b.writerIndex();
+
+            /* consider a 2% less than writesize limit check on smr entries. Checkpoint entry
+             * adds more metadata to the actual size of SMR entries. Its a safeguard against the
+             * smr entries amounting to the actual boundary limit. */
+            if (currentBatchSize == batchSize || numBytesPerCheckpointEntry >= maxWriteSizeLimit) {
```

Comment:
Q. If the first comparison is there only for 'testing' purposes (like suggested in the comments) aren't we then limiting production code to this? cause if that condition is met before the size limit we won't be taking advantage of this change... I don't see any issue as that was how it was working before, I'm just asking if we have thought it through and we'd prefer to adjust our tests cause it's not that relevant or if its ok..

Reply:
It would be only >? cause if it's equal to we could still add it to this batch, right?

Reply:
I thought that keeping the old check (number of entries) and adding a new check (numBytes) is safer than replacing the old one. I'm not sure if removing the batchSize limit would harm / impose side-effects on somewhere else in Corfu (just from my guess). I do totally agree that removing batchSize check will improve performance.

Reply:
@Maithem As per Anny's suggestion, removing the limit of max number of entries would improve the checkpoint performance, could you give your thoughts on this? From the fact that WriteSizeException has not been triggered until very recently, the total numBytes must be much lower than 25MB under the old 50 batch size limit, which means it has been underusing the stream capacity.  If having only one limit on numBytes does not harm anywhere else in Corfu, I think it would be an optimization for the slow checkpointing process today.

Reply:
I think it can improve performance, but the current state is probably going to decrease throughput. Imagine the case where you have a batch of 50 entries and that each entry is 2500000 bytes, then this algorithm will generate a batch size of 10, but the existing algorithm would generate a batch size of 50. Assuming a compression rate of 5x, this change will decrease throughput in comparison to what's on master. 

What is missing here is that the serialized size is accounted for before compression and that is not accurate.